turbopack/crates/turbopack-core/src/module_graph/mod.rs

```diff
@@ -583,6 +583,103 @@ impl SingleModuleGraph {
 
         Ok(())
     }
+
+    pub fn traverse_cycles<'l>(
+        &'l self,
+        edge_filter: impl Fn(&'l ChunkingType) -> bool,
+        mut visit_cycle: impl FnMut(&[&'l SingleModuleGraphModuleNode]),
+    ) {
+        // see https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+        // but iteratively instead of recursively
+
+        #[derive(Clone)]
+        struct NodeState {
+            index: u32,
+            lowlink: u32,
+            on_stack: bool,
+        }
+        enum VisitStep {
+            UnvisitedNode(NodeIndex),
+            EdgeAfterVisit { parent: NodeIndex, child: NodeIndex },
+            AfterVisit(NodeIndex),
+        }
+        let mut node_states = vec![None; self.graph.node_bound()];
+        let mut stack = Vec::new();
+        let mut visit_stack = Vec::new();
+        let mut index = 0;
+        let mut scc = Vec::new();
+        for initial_index in self.graph.node_indices() {
+            // Skip over already visited nodes
+            if node_states[initial_index.index()].is_some() {
+                continue;
+            }
+            visit_stack.push(VisitStep::UnvisitedNode(initial_index));
+            while let Some(step) = visit_stack.pop() {
+                match step {
+                    VisitStep::UnvisitedNode(node) => {
+                        node_states[node.index()] = Some(NodeState {
+                            index,
+                            lowlink: index,
+                            on_stack: true,
+                        });
+                        index += 1;
+                        stack.push(node);
+                        visit_stack.push(VisitStep::AfterVisit(node));
+                        let mut neighbors = self.graph.neighbors(node).detach();
+                        while let Some((edge, succ)) = neighbors.next(&self.graph) {
+                            let edge_weight = self.graph.edge_weight(edge).unwrap();
+                            if !edge_filter(edge_weight) {
```

Comment:
The edge_filter should only apply to the cycles though and not cause the traversal to end? Otherwise you'd never discover cycles somewhere behind a filtered edge.